Make named pipe security configurable by security descriptor string

Breakwell opened the pull request:
### Description
Named pipes on Windows are vulnerable to privilege escalation through named pipe impersonation, to protect against this in Thrift it should allow the security of a pipe to be configured.

### Implementation
Changed setting the pipe permissions to "everyone" to now be configured by security string this will default to "everyone" but can also by set in the constructor of a TPipeServer.

kainjow commented:
you should probably make a Jira, this isn't a trivial change.

kainjow commented on lib/cpp/src/thrift/transport/TPipeServer.cpp:
```diff
@@ -323,73 +344,47 @@ void TPipeServer::close() {
-  LocalFree(acl);
-  FreeSid(everyone_sid);
+  if (psd)
+    LocalFree(psd);
```
can we use an RAII wrapper instead?

Breakwell replied:
I feel it's a bit overkill to add a RAII considering we only free once.

kainjow commented on lib/cpp/src/thrift/transport/TPipeServer.h:
```diff
@@ -76,6 +85,7 @@ class TPipeServer : public TServerTransport {
   bool getAnonymous();
   void setAnonymous(bool anon);
   void setMaxConnections(uint32_t maxconnections);
+  void setSecurityDescriptor(std::string securityDescriptor);
```
make const reference.

Breakwell replied:
Done.

kainjow commented:
I would suggest backing out the extra style changes to keep the diff minimal and relevant.

jeking3 commented on lib/cpp/src/thrift/transport/TPipeServer.cpp:
```diff
@@ -95,9 +96,15 @@ class TAnonPipeServer : public TPipeServerImpl {
+  TNamedPipeServer(const std::string& pipename,
+                   uint32_t bufsize,
+                   uint32_t maxconnections,
+                   const std::string& securityDescriptor)
```
> to protect against this in Thrift it should allow the security of a pipe to be configured

This forces the issue, and is a breaking change.  It would be better to introduce this functionality while still allowing existing code to compile, and deprecate the old constructors, and then remove them in the following release if we really think it is necessary to force a securityDescriptor in all use cases (I'm not convinced yet).

Breakwell replied:
I agree that this shouldn't be a breaking a change and shouldn't enforce the use of a security descriptor. TNamedPipeServer is internal to TPipeServer. I have given TPipeServer an additional constructor that now takes a security descriptor as a parameter. Now when TPipeServer is constructed unless a security descriptor is provided it has the security descriptor set to "Everyone" to maintain the previous behavior . As TNamedPipeServer  is only in the cpp I don't see how this is a breaking change.

kainjow commented on lib/cpp/src/thrift/transport/TPipeServer.cpp:
```diff
@@ -400,13 +395,12 @@ bool TAnonPipeServer::createAnonPipe() {
   }
-  if (!SetSecurityDescriptorDacl(&sd, true, NULL, false))
-  {
+  if (!SetSecurityDescriptorDacl(&sd, true, NULL, false)) {
```
nit: remove brace line change.

Breakwell replied:
This is now inline with the clang format file in the repo. As it's a step in the right direction and doesn't make the PR unclear I think it's fine to leave the formatting change as is.

kainjow commented on lib/cpp/src/thrift/transport/TPipeServer.cpp:
```diff
@@ -400,13 +395,12 @@ bool TAnonPipeServer::createAnonPipe() {
-  if (!InitializeSecurityDescriptor(&sd, SECURITY_DESCRIPTOR_REVISION))
-  {
-    GlobalOutput.perror("TPipeServer InitializeSecurityDescriptor (anon) failed, GLE=", GetLastError());
+  if (!InitializeSecurityDescriptor(&sd, SECURITY_DESCRIPTOR_REVISION)) {
```
nit: remove unneeded formatting changes.

kainjow commented on lib/cpp/src/thrift/transport/TPipeServer.cpp:
```diff
@@ -323,73 +344,47 @@ void TPipeServer::close() {
-    throw TTransportException(TTransportException::NOT_OPEN,
-                              "TCreateNamedPipe() failed",
-                lastError);
+    throw TTransportException(TTransportException::NOT_OPEN, "TCreateNamedPipe() failed",
```
nit: remove unneeded formatting changes

Jens-G commented:
Not quite sure .... what is tghe status of this PR? Are we done or are there any pending change requests?

Breakwell commented:
> Not quite sure .... what is tghe status of this PR? Are we done or are there any pending change requests?

I have addressed the comments for clarity. I am happy with the changes unless anyone feels particularly strongly I think it's fine as is.
